### ISSUE TITLE
UninitializedPropertyAccessException when used in Java project

### ISSUE THREAD

**Issue by koostamas** [diogobernardino/WilliamChart#251]
I tried to put the chart from the example app into my own app, but it crashes as soon as I open the fragment that the chart is on. The exception: `
kotlin.UninitializedPropertyAccessException: lateinit property data has not been initialized
        at com.db.williamchart.renderer.BarChartRenderer.preDraw(BarChartRenderer.kt:67)
        at com.db.williamchart.view.BarChartView$$special$$inlined$doOnPreDraw$1.onPreDraw(View.kt:348)
        at android.view.ViewTreeObserver.dispatchOnPreDraw(ViewTreeObserver.java:977)
        at android.view.ViewRootImpl.performTraversals(ViewRootImpl.java:2337)
        at android.view.ViewRootImpl.doTraversal(ViewRootImpl.java:1386)
        at android.view.ViewRootImpl$TraversalRunnable.run(ViewRootImpl.java:6733)
        at android.view.Choreographer$CallbackRecord.run(Choreographer.java:911)
        at android.view.Choreographer.doCallbacks(Choreographer.java:723)
        at android.view.Choreographer.doFrame(Choreographer.java:658)
        at android.view.Choreographer$FrameDisplayEventReceiver.run(Choreographer.java:897)
        at android.os.Handler.handleCallback(Handler.java:789)
        at android.os.Handler.dispatchMessage(Handler.java:98)
        at android.os.Looper.loop(Looper.java:164)
        at android.app.ActivityThread.main(ActivityThread.java:6541)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.Zygote$MethodAndArgsCaller.run(Zygote.java:240)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:767)`
My relevant code:
`@Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        LinkedHashMap<String, Float> lineMap = new LinkedHashMap<>();
        lineMap.put("label1", 5f);
        lineMap.put("label2", 4.5f);
        lineMap.put("label3", 4.7f);
        lineMap.put("label4", 3.5f);
        lineMap.put("label5", 3.6f);
        lineMap.put("label6", 7.5f);
        lineMap.put("label7", 7.5f);
        lineMap.put("label8", 10f);
        lineMap.put("label9", 5f);
        lineMap.put("label10", 6.5f);
        lineMap.put("label11", 3f);
        lineMap.put("label12", 4f);
        int[] fillColors = {Color.parseColor("#81FFFFFF"), Color.TRANSPARENT};
        binding.lineChart.setGradientFillColors(fillColors);
        binding.lineChart.getAnimation().setDuration(1000);
        binding.lineChart.animate(lineMap);
    }`

**Comment by ardevd** [comment]
I get the same issue here too

**Comment by mathmerizing** [comment]
I have the same issue as well

**Comment by abeemukthees** [comment]
Got the same issue in Kotlin too
```
2019-09-18 23:12:04.279 8092-8099/? E/illiamchartdem: Failed to send DDMS packet REAQ to debugger (-1 of 20): Broken pipe
2019-09-18 23:12:04.678 8092-8092/? E/AndroidRuntime: FATAL EXCEPTION: main
    Process: com.db.williamchartdemo, PID: 8092
    kotlin.UninitializedPropertyAccessException: lateinit property data has not been initialized
        at com.db.williamchart.renderer.LineChartRenderer.preDraw(LineChartRenderer.kt:57)
        at com.db.williamchart.view.LineChartView$$special$$inlined$doOnPreDraw$1.run(View.kt:357)
        at androidx.core.view.OneShotPreDrawListener.onPreDraw(OneShotPreDrawListener.java:71)
        at android.view.ViewTreeObserver.dispatchOnPreDraw(ViewTreeObserver.java:977)
        at android.view.ViewRootImpl.performTraversals(ViewRootImpl.java:2474)
        at android.view.ViewRootImpl.doTraversal(ViewRootImpl.java:1460)
        at android.view.ViewRootImpl$TraversalRunnable.run(ViewRootImpl.java:7183)
        at android.view.Choreographer$CallbackRecord.run(Choreographer.java:949)
        at android.view.Choreographer.doCallbacks(Choreographer.java:761)
        at android.view.Choreographer.doFrame(Choreographer.java:696)
        at android.view.Choreographer$FrameDisplayEventReceiver.run(Choreographer.java:935)
        at android.os.Handler.handleCallback(Handler.java:873)
        at android.os.Handler.dispatchMessage(Handler.java:99)
        at android.os.Looper.loop(Looper.java:193)
        at android.app.ActivityThread.main(ActivityThread.java:6669)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:493)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:858)

```

**Comment by diogobernardino** [comment]
Looks like the preDraw is called before the `show()` or `animate()`, so your data isn't yet initialized. Wondering if it has anything to do with data binding layout, it's something different from Demo app. 
@ardevd @abeemukthees could you also both provide the chart configuration your doing (e.g. your data, chart configuration)?

**Comment by diogobernardino** [comment]
@koostamas your codes tells me you're configuring a line chart but your crash reports a bar chart. Do you also have a bar chart defined in the layout? If so, are you providing it values?

**Comment by abeemukthees** [comment]
All I did was just commented on one line.
```
override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        return inflater.inflate(R.layout.demo_fragment, container, false)
    }

    override fun onViewCreated(view: View, saveInstanceState: Bundle?) {

        lineChart.gradientFillColors =
            intArrayOf(
                Color.parseColor("#81FFFFFF"),
                Color.TRANSPARENT
            )

        lineChart.animation.duration = 1000
        //lineChart.animate(lineSet)



        horizontalBarChart.animation.duration = 1000
        horizontalBarChart.animate(horizontalBarSet)
    }
```
Error:
```
:43.696 4848-4848/com.db.williamchartdemo D/AndroidRuntime: Shutting down VM
2019-09-19 22:15:43.719 4848-4848/com.db.williamchartdemo E/AndroidRuntime: FATAL EXCEPTION: main
    Process: com.db.williamchartdemo, PID: 4848
    kotlin.UninitializedPropertyAccessException: lateinit property data has not been initialized
        at com.db.williamchart.renderer.LineChartRenderer.preDraw(LineChartRenderer.kt:57)
        at com.db.williamchart.view.LineChartView$$special$$inlined$doOnPreDraw$1.run(View.kt:357)
        at androidx.core.view.OneShotPreDrawListener.onPreDraw(OneShotPreDrawListener.java:71)
        at android.view.ViewTreeObserver.dispatchOnPreDraw(ViewTreeObserver.java:977)
        at android.view.ViewRootImpl.performTraversals(ViewRootImpl.java:2474)
        at android.view.ViewRootImpl.doTraversal(ViewRootImpl.java:1460)
        at android.view.ViewRootImpl$TraversalRunnable.run(ViewRootImpl.java:7183)
        at android.view.Choreographer$CallbackRecord.run(Choreographer.java:949)
        at android.view.Choreographer.doCallbacks(Choreographer.java:761)
        at android.view.Choreographer.doFrame(Choreographer.java:696)
        at android.view.Choreographer$FrameDisplayEventReceiver.run(Choreographer.java:935)
        at android.os.Handler.handleCallback(Handler.java:873)
        at android.os.Handler.dispatchMessage(Handler.java:99)
        at android.os.Looper.loop(Looper.java:193)
        at android.app.ActivityThread.main(ActivityThread.java:6669)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:493)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:858)
```

**Comment by diogobernardino** [comment]
You have a line chart defined in the xml but afterwards you provide no datapoints to be displayed (meaning that you do not call the `animate()`). I do that on purpose as I assumed it wouldn't make sense to have a chart without providing data, but now I see it can actually bring confusion. I need to re-think a better strategy.

If all I said above it's correct you should overcome the crash by either not comment `animate()` or removing the `LineChartView` completely from your layout.

**Comment by abeemukthees** [comment]
How can call `animate(lineSet)` when I'm still waiting for the data to load from network.
Also please provide support to `animate(newLineSet)` when the data has changed.

**Comment by diogobernardino** [comment]
Agree. I will fix it.

As for animating a chart from an old set to a new set, it's definitely a good improvement, however not sure when I will be able to ship it.

**Comment by abeemukthees** [comment]
Thank u, U can look at https://github.com/robinhood/spark for an idea.

**Comment by diogobernardino** [comment]
I just published `3.1.1`, you may want to give it a try.

> Thank u, U can look at https://github.com/robinhood/spark for an idea.

Had no idea they had it open source, I will definitely have a look. Thanks.

**Comment by abeemukthees** [comment]
> I just published `3.1.1`, you may want to give it a try.
> 
> > Thank u, U can look at https://github.com/robinhood/spark for an idea.
> 
> Had no idea they had it open source, I will definitely have a look. Thanks.

It's not crashing now. Thanks.

If you replicate that morphing animation from `Spark` to `LineChartView` for dataSet changes, that would be awesome.
